effectful/internals/base_impl.py

```diff
@@ -86,25 +74,38 @@ def __free_rule__(self, *args: Q.args, **kwargs: Q.kwargs) -> "Expr[V]":
             else:
                 unscoped_args.add(param_name)
 
+        if not bound_vars:  # fast path for no bound variables
+            return (
+                tuple(frozenset() for _ in bound_sig.args),
+                {k: frozenset() for k in bound_sig.kwargs},
+            )
+
         # TODO replace this temporary check with more general scope level propagation
-        if bound_vars:
-            min_scope = min(bound_vars.keys(), default=0)
-            scoped_args[min_scope] |= unscoped_args
-            max_scope = max(bound_vars.keys(), default=0)
-            assert all(s in bound_vars or s > max_scope for s in scoped_args.keys())
+        min_scope = min(bound_vars.keys(), default=0)
+        scoped_args[min_scope] |= unscoped_args
+        max_scope = max(bound_vars.keys(), default=0)
+        assert all(s in bound_vars or s > max_scope for s in scoped_args.keys())
 
         # recursively rename bound variables from innermost to outermost scope
```

Comment:
Why are we renaming bound variables? If this rule returns the free variables in each position in the arguments list, I would expect bound variables to be removed from these sets.

Reply:
This comment is obsolete after the changes in this PR and should be removed.

Reply:
All of this is removed in #208